### DISCUSSION TITLE
Add out of tree sources to an existing zephyr library (drivers__ethernet)

### DISCUSSION THREAD

**Discussion by ofirshe** [zephyrproject-rtos/zephyr#53283] [Q&A]
Hey, 

I developed a low level network driver for out custom board and I would like to add my sources to an existing library drivers__ethernet. 
Is it possible ? The only possible option I thought about is to clone a local copy of zpehyr and add my driver inside but I want to avoid this option.

Currently I got the message
 CMake Warning at zephyr/CMakeLists.txt:824 (message):
  No SOURCES given to Zephyr library: drivers__ethernet

  Excluding target from build.

Thanks, 
Ofir.

**Comment by saacifuentesmu-titoma** [comment]
Hey ofirshe,

I'm running into the same problem with a custom LCD driver. Did you find any acceptable solution?

**Comment by yashi** [comment]
It _is_ possible.  I've done that.

The message says that no driver from `zephyr/drivers/ethernet` is selected and removing the directory.  It doesn't say the build system is removing your out-of-tree driver.

To inhibit the message, there is are cmake function, which is not installed in `zephyr/drivers/ethernet`, but in `zephyr/drivers/sensor/CMakeLists.txt`.

```cmake
zephyr_library_property(ALLOW_EMPTY TRUE)
```

**Comment by henrikbrixandersen** [comment]
For out-of-tree drivers, you can use `zephyr_library_amend()` as documented here: https://github.com/zephyrproject-rtos/zephyr/blob/1e471eb2a79fe86c854455df16744f2a20768492/cmake/modules/extensions.cmake#L431

Beware that your directory structure must match that of Zephyr (i.e. use `drivers/ethernet/` in this case).

**Reply by josuah** [reply]
I knew this file existence, I even read it, but somehow I missed that one.
Glad to learn about it!

**Reply by hraftery** [reply]
In my case the OOT driver is in a folder within my application project. `zephyr_library_amend()` fails because it's not within a standalone Zephyr  module. Rather than re-architect, I just re-implemented the macro like so:

```
zephyr_library_get_current_dir_lib_name(${OOT_BASE} lib_name)
set(ZEPHYR_CURRENT_LIBRARY ${lib_name})
```

where `OOT_BASE` is set in the same style as `ZEPHYR_BASE` (ie. an absolute path) to my OOT base folder, like so:

```
set(OOT_BASE ${CMAKE_CURRENT_SOURCE_DIR}/src/zephyr)
```

Very cool result. My OOT folder structure mimics Zephyr's, but with just my extensions. I do no explicit library linking, and just kick off the CMake process with `add_subdirectory(${OOT_BASE})` in my project's `CMakeLists.txt` file. All seems to build and link nicely.

**Comment by saacifuentesmu-titoma** [comment]
Thank you guys. 

The problem was that the custom driver Kconfig files were not included int he build I think.

Made it work some days ago by compiling my project inside a zephyr workspace.
IE, created and updated the workspace as explained in [zephyrproject-rtos/example-application.](https://github.com/zephyrproject-rtos/example-application) I think this should be done after uploading the west manifest to the cloud.

All the custom driver Kconfig related files were included after that, and the LCD driver was included even if that "Excluding target from build" was printed out.
